action/protocol/staking/handlers.go

```diff
@@ -158,23 +161,196 @@ func (p *Protocol) handleWithdrawStake(ctx context.Context, act *action.Withdraw
-	// TODO
-	return nil, nil
+	candidate := p.inMemCandidates.GetByName(act.Candidate())
+	if candidate == nil {
+		return nil, errors.Wrap(ErrInvalidCanName, "cannot find candidate in candidate center")
+	}
+
+	_, gasFee, err := fetchCaller(ctx, sm, big.NewInt(0))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch caller")
+	}
+
+	bucket, err := p.fetchBucket(ctx, sm, act.BucketIndex(), true, false)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch bucket")
+	}
+
+	prevCandidate := p.inMemCandidates.GetByOwner(bucket.Candidate)
+	if prevCandidate == nil {
+		return nil, errors.Wrap(ErrInvalidOwner, "cannot find candidate in candidate center")
+	}
+
+	// update bucket index
+	if err := delCandidateIndex(sm, bucket.Candidate, act.BucketIndex()); err != nil {
+		return nil, errors.Wrapf(err, "failed to delete candidate bucket index for candidate %s", bucket.Candidate.String())
+	}
+	if err := putCandidateIndex(sm, candidate.Owner, act.BucketIndex()); err != nil {
+		return nil, errors.Wrapf(err, "failed to put candidate bucket index for candidate %s", candidate.Owner.String())
+	}
+	// update bucket
+	bucket.Candidate = candidate.Owner
+	if err := updateBucket(sm, act.BucketIndex(), bucket); err != nil {
+		return nil, errors.Wrapf(err, "failed to update bucket for voter %s", bucket.Owner)
+	}
+
+	weightedVotes := p.calculateVoteWeight(bucket, false)
+
+	// update previous candidate
+	if err := prevCandidate.SubVote(weightedVotes); err != nil {
+		return nil, errors.Wrapf(err, "failed to subtract vote for previous candidate %s", prevCandidate.Owner.String())
+	}
+	if err := putCandidate(sm, prevCandidate); err != nil {
+		return nil, errors.Wrapf(err, "failed to put state of previous candidate %s", prevCandidate.Owner.String())
+	}
+
+	// update current candidate
+	if err := candidate.AddVote(weightedVotes); err != nil {
+		return nil, errors.Wrapf(err, "failed to add vote for candidate %s", candidate.Owner.String())
+	}
+	if err := putCandidate(sm, candidate); err != nil {
+		return nil, errors.Wrapf(err, "failed to put state of candidate %s", candidate.Owner.String())
+	}
+
+	receipt, err := p.settleAction(ctx, sm, gasFee)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to settle action")
+	}
+	p.inMemCandidates.Delete(prevCandidate.Owner)
+	p.inMemCandidates.Put(prevCandidate)
+	p.inMemCandidates.Delete(candidate.Owner)
+	p.inMemCandidates.Put(candidate)
+	return receipt, nil
 }
 
 func (p *Protocol) handleTransferStake(ctx context.Context, act *action.TransferStake, sm protocol.StateManager) (*action.Receipt, error) {
-	// TODO
-	return nil, nil
+	_, gasFee, err := fetchCaller(ctx, sm, big.NewInt(0))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch caller")
+	}
+
+	bucket, err := p.fetchBucket(ctx, sm, act.BucketIndex(), true, false)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch bucket")
+	}
+
+	voterAddr, err := address.FromString(act.VoterAddress())
```

Comment:
i think act.VoterAddress() should be address.Address
i.e., we discuss that action.TransferStake should handle this conversion from string to address.Address